torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,240 @@
+                    )
+                else:
+                    # cache only once in the non-causal case
+                    self.cache = dict(k=k.clone(), v=v.clone())
+            if self.causal:
+                idx = (
+                    slice(None, None),
+                    slice(None, None),
+                    *[slice(i, i + 1) for i in decode_idx],
+                )
+                self.cache["k"][idx] = k
+                self.cache["v"][idx] = v
+            k, v = self.cache["k"], self.cache["v"]
+
+        a = self.attn(q, k, v, decode_step, decode_idx)
+        a = self._combine_multihead(a)
+        a = self.fc(a)
+
+        return a
+
+
+class FullAttention(nn.Module):
+    """Computes attention over the entire flattened input.
+
+    Args:
+        shape (Tuple): shape of input data (d1 x ... x dn)
+        causal (bool): use causal attention or not
+        attn_dropout (float): probability of dropout after softmax
+
+    Inputs:
+        q, k, v (Tensor): a [b, d1, ..., dn, c] tensor or
+                          a [b, 1, ..., 1, c] tensor if decode_step is not None
+
+    """
+
+    def __init__(
+        self, shape: Tuple, causal: bool = False, attn_dropout: float = 0.0
+    ) -> None:
+        super().__init__()
+        self.causal = causal
+        self.attn_dropout = attn_dropout
+
+        if self.causal:
+            seq_len = int(torch.prod(torch.tensor(shape)).item())
+            self.register_buffer("mask", torch.tril(torch.ones(seq_len, seq_len)))
+
+    def forward(
+        self, q: Tensor, k: Tensor, v: Tensor, decode_step=None, decode_idx=None
+    ) -> Tensor:
+        mask = torch.Tensor(self.mask) if self.causal else None
+        if decode_step is not None and mask is not None:
+            mask = mask[[decode_step]]
+
+        elif mask is not None and q.size(2) < mask.size(0):
+            mask = mask[range(q.size(2)), :][:, range(q.size(2))]
+
+        old_shape = q.shape[2:-1]
+        q = q.flatten(start_dim=2, end_dim=-2)
+        k = k.flatten(start_dim=2, end_dim=-2)
+        v = v.flatten(start_dim=2, end_dim=-2)
+
+        out = scaled_dot_product_attention(
+            q, k, v, mask=mask, attn_dropout=self.attn_dropout, training=self.training
+        )
+
+        return out.unflatten(2, old_shape)
+
+
+class AxialAttention(nn.Module):
+    """Computes attention over a single axis of the input. Other dims are flattened
+    into the batch dimension.
+
+    Args:
+        axial_dim (int): dimension to compute attention on, index by input dimensions
+                         (i.e., 0 for first input dimension, 1 for second)
+
+    Inputs:
+        q, k, v (Tensor): a [b, h, d1, ..., dn, c] tensor or
+                          a [b, h, 1, ..., 1, c] tensor if decode_step is not None
+
+    """
+
+    def __init__(self, axial_dim: int) -> None:
+        super().__init__()
+        self.axial_dim = axial_dim + 2  # account for batch, head
```

Comment:
imo it's confusing because first case ignores batch, head, hidden dim but the negative case does not, at least in MUGEN code. I didn't think it would be necessary to support this. Supporting negative indexing where it ignores the other dims might be nice, I could add that.

Reply:
if we want to specify the second axis of a 3d input, we should be able to use 1 or -2. If the input tensor is `b,h,d1,d2,d3,c`, then 1 += 2 maps to 3 which is correct. 2 + (3 + 1 + -2) maps to 4 which is not correct... the user would have to use -3 to account for channel dimension, but in the positive case 1 does not have to account for any extra dimension.